### PR TITLE
Add PyTorch dataset and dataloader classes for TemporalData

### PR DESCRIPTION
**Pull request by yulonglin**
## What
This adds native PyTorch `DataLoader` support for `TemporalData`, which is used for dynamic graph datasets. This is based on the idea mentioned here: #2567 

## Why
By allowing dynamic graph events in `TemporalData` format to be loaded via the new `TemporalDataLoader`, we can inherit the functionality of PyTorch `DataLoader`. For example, this is useful for:
- Interfacing with other libraries such as PyTorch Lightning, where a PyTorch `DataLoader` is expected
- Shuffling data loading using the `shuffle` argument when defining static baselines where time information is not important
- Parallelising data loading using the `num_workers` argument, in particular when there is preprocessing

Note: Currently, `TemporalData` has a `seq_batches()` method that generates slices of a tensor, batch by batch. In the case where we have a small dataset that we just need to load sequentially without preprocessing and don't need a `DataLoader`-like API, `seq_batches()` is faster. This is because using PyTorch `DataLoader` means splitting the data into individual training samples before collating them, but `seq_batches` skips the splitting and collating steps. This is similar to PyTorch and PyTorch Geometric `DataLoader` classes.

## How
A `TemporalDataset` is created that wraps around `TemporalData` to create a PyTorch `Dataset` object. Individual samples are collated by the `TemporalDataLoader` using a special collater for `TemporalData` objects.  

## Testing
Tests were implemented to check for the shapes and values of inputs and outputs to `TemporalDataLoader`. The TGN-attn implementation in `examples/` was modified by replacing the `seq_batches()` method with `TemporalDataLoader` and it produced similar results.

### CODE REVIEW

**Comment by Padarn** [comment]
hey @yulonglin and @rusty1s  - I was looking at the requirement for this from https://github.com/pyg-team/pytorch_geometric/issues/3230, what are your plans for this PR?

**Comment by rusty1s** [comment]
I'm in favor of re-using the `torch_geometric.loader.DataLoader` for mini-batching of temporal graphs as well. This should be (hopefully) supported out-of-the-box once `TemporalData` inherits from `BaseData`.

**Comment by Padarn** [comment]
Yeah that makes a lot of sense to me. Will revisit if anything is needed once that is addressed from https://github.com/pyg-team/pytorch_geometric/issues/3230

**Comment by rusty1s** [comment]
Closing this in favor of https://github.com/pyg-team/pytorch_geometric/pull/3985. Thank you!
